torch/nn/modules/transformer.py

```diff
@@ -375,6 +375,81 @@ def forward(self, tgt, memory, tgt_mask=None, memory_mask=None,
         return tgt
 
 
+class MultiheadAttentionInProjection(Module):
+    def __init__(self, in_embed_dim, num_heads, out_embed_dim=None):
+        super(MultiheadAttentionInProjection, self).__init__()
+        self.in_embed_dim = in_embed_dim
+        if not out_embed_dim:
+            self.out_embed_dim = in_embed_dim
+        else:
+            self.out_embed_dim = out_embed_dim
+        self.num_heads = num_heads
+        self.fc = Linear(self.in_embed_dim, self.out_embed_dim)
+
+    def forward(self, query):
+        seq, bsz, embed_dim = query.size(0), query.size(1), query.size(2)
+        head_dim = embed_dim // self.num_heads
+        assert head_dim * self.num_heads == embed_dim, \
+            "embed_dim must be divisible by num_heads"
+
+        q = self.fc(query)
+        q = torch.reshape(q, (seq, bsz * self.num_heads, head_dim))
+        return q.transpose(0, 1)
+
+
+class ScaledDotProduct(Module):
+    def __init__(self, dropout=0.0):
+        super(ScaledDotProduct, self).__init__()
+        self.dropout = dropout
+
+    def forward(self, query, key, value, attn_mask=None):
```

Comment:
I think this already supports broadcasting across query, key and value. This is worth verifying and documenting in my opinion.